lib/pkcs11_cat.c

```diff
@@ -268,15 +265,11 @@ func_rc pkcs11_cat_object_with_label(pkcs11Context *p11Context, char *label, int
 				    goto key_dsa_error;
 				}
 
-				if(dsa->p !=NULL ) { BN_free(dsa->p); dsa->p = NULL; }
-				if(dsa->q !=NULL ) { BN_free(dsa->q); dsa->q = NULL; }
-				if(dsa->g !=NULL ) { BN_free(dsa->g); dsa->g = NULL; }
-				if(dsa->pub_key !=NULL ) { BN_free(dsa->pub_key); dsa->pub_key = NULL; }
-
-				dsa->p = bn_prime;        bn_prime = NULL;    /* forget, moved to dsa */
-				dsa->q = bn_subprime;     bn_subprime = NULL; /* forget, moved to dsa */
-				dsa->g = bn_base;         bn_base = NULL;     /* forget, moved to dsa */
-				dsa->pub_key = bn_pubkey; bn_pubkey = NULL;   /* forget, moved to dsa */
+        DSA_set0_pqg(dsa, bn_prime, bn_subprime, bn_base);
+				bn_prime = NULL;    /* forget, moved to dsa */
+				bn_subprime = NULL; /* forget, moved to dsa */
+				bn_base = NULL;     /* forget, moved to dsa */
+				DSA_set0_key(dsa, bn_pubkey, NULL);
```

Comment:
`bn_pubkey = NULL;` is missing after this line, resulting in a core dump:

Tests are showing that a core dump occurs as a result of object being freed twice.

Please add the missing assignment.

Reply:
I see similar later too

Reply:
> I see similar later too

Ian, would you be kind enough to fix the one above and the other you know of, and commit on your remote branch, so it gets reflected in this PR?

Thanks a lot for your work!

Reply:
I think i've sorted it out. :)